### PR TITLE
Support for Diameter protocol layer (base RFC6733 + extensions)

### PR DESCRIPTION
**Pull request by gpotter2**
Original code: https://bitbucket.org/secdev/scapy/pull-requests/109/diameter-layer-support-v10

This is a fix for the very old https://github.com/secdev/scapy/pull/160, working with current scapy.

I haven't reviewed the original entire code, but fixed failures, tabs (tabs are not supported under python 3+).

### CODE REVIEW

File: scapy/layers/sctp.py [hunk 1]
```diff
@@ -433,6 +530,10 @@ class SCTPChunkShutdownComplete(_SCTPChunkGuessPayload, Packet):
                     ShortField("len", 4),
                     ]
 
-bind_layers( IP,           SCTP,          proto=IPPROTO_SCTP)
+bind_layers ( IP,           SCTP,          proto=IPPROTO_SCTP)
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Please avoid unneeded changes (plus this one is wrong: no space needed here.

File: scapy/layers/sctp.py [hunk 2]
```diff
@@ -16,6 +16,10 @@
 from scapy.layers.inet import IP
 from scapy.layers.inet6 import IP6Field
 from scapy.layers.inet6 import IPv6
+try:
+  from scapy.contrib.diameter import DiamG
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Wrong indent here, I think.

File: scapy/layers/sctp.py [hunk 3]
```diff
@@ -178,7 +182,7 @@ def sctp_checksum(buf):
 
 # Dummy class to guess payload type (variable parameters)
 class _SCTPChunkGuessPayload:
-    def default_payload_class(self,p):
+    def default_payload_class(self,p):
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Wrong and unneeded change here.

File: scapy/layers/sctp.py (hunk 2 above)
Comment (p-l-) [inline scapy/layers/sctp.py]: This should not happen: if `diameter` contribution is loaded, then it should add the layer binding. But we don't want a layer to depend on a contrib.

File: scapy/layers/sctp.py [hunk 4]
```diff
@@ -294,21 +298,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
 
 ############## SCTP Chunks
 
+# Dictionary taken from: http://www.iana.org/assignments/sctp-parameters/sctp-parameters.xhtml
+sctp_payload_protocol_identifiers = {
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Please conform to PEP-8 for new code.

Comment (gpotter2) [inline scapy/layers/sctp.py]: Should be better

**Comment by guedou** [comment]
@p-l- do you agree with the changes ?

**Comment by guedou** [comment]
@gpotter2 what the current coverage rate of diameter.py ?

**Comment by gpotter2** [comment]
@guedou 
> 91.76%

**Comment by guedou** [comment]
@p-l- do you agree with the changes ?

**Comment by gpotter2** [comment]
Just rebased to fix coverage.

File: scapy/layers/sctp.py (hunk 4 above)
Comment (p-l-) [inline scapy/layers/sctp.py]: Constants should be uppercase I think.

Comment (gpotter2) [inline scapy/layers/sctp.py]: Should be better now

**Comment by gpotter2** [comment]
@p-l- Hi any new on this ?

File: scapy/layers/sctp.py [hunk 5]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
 class SCTPChunkData(_SCTPChunkGuessPayload, Packet):
+# TODO : add a padding function in post build if this layer is used to generate SCTP chunk data
     fields_desc = [ ByteEnumField("type", 0, sctpchunktypes),
                     BitField("reserved", None, 4),
                     BitField("delay_sack", 0, 1),
                     BitField("unordered", 0, 1),
                     BitField("beginning", 0, 1),
                     BitField("ending", 0, 1),
-                    FieldLenField("len", None, length_of="data", adjust = lambda pkt,x:x+16),
+                    ShortField("len", None),
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Could you explain why you make this change?

File: scapy/layers/sctp.py [hunk 6]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
                     XIntField("tsn", None),
                     XShortField("stream_id", None),
                     XShortField("stream_seq", None),
-                    XIntField("proto_id", None),
+                    ProtoIdField("proto_id", None, SCTP_PAYLOAD_PROTOCOL_INDENTIFIERS),
```
Comment (p-l-) [inline scapy/layers/sctp.py]: You should use an `IntEnumField()` here.

File: scapy/contrib/diameter.py [hunk 7]
```diff
@@ -0,0 +1,2596 @@
+#!/usr/bin/env python
```
Comment (p-l-) [inline scapy/contrib/diameter.py]: This file is far from being PEP-8 compliant. We cannot merge it as it.

File: scapy/contrib/diameter.py [hunk 8]
```diff
@@ -0,0 +1,2596 @@
...
+
+class DRFlags (FlagsField):
+    def i2repr (self, pkt, x):
+        if x == None:
+            return "None"
+        res = hex(int(x))
+        r = ''
+        cmdt = (x & 128) and ' Request' or ' Answer'
+        if x & 15:  # Check if reserved bits are used
+                nb = 8 ; offset = 0
```
Comment (p-l-) [inline scapy/contrib/diameter.py]: This should be two lines. `;` are forbidden! ;-)

File: scapy/layers/sctp.py [hunk 9]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
+      if self.proto_id in protoClassname: # Use this value to set the ChunckData decoding class
+        className = protoClassname [self.proto_id]
+        return globals().get(className, conf.raw_layer)        
+      else: # guess payload class after having set dport and sport fields        
+        ul = self.underlayer  # Not always the SCTP layer (may be the SACK one)
+        while True:   # Get the SCTP layer   
+          tul = type(ul)
+          if tul == SCTP:
+            break
+          elif tul == type(None): # Should not reach this point but just in case ...
```
Comment (p-l-) [inline scapy/layers/sctp.py]: This test should be written as `elif ul is None`.

File: scapy/layers/sctp.py [hunk 10]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
+        47  : "DiamG",    # Diameter
+      }
+      if self.proto_id in protoClassname: # Use this value to set the ChunckData decoding class
+        className = protoClassname [self.proto_id]
+        return globals().get(className, conf.raw_layer)        
+      else: # guess payload class after having set dport and sport fields        
+        ul = self.underlayer  # Not always the SCTP layer (may be the SACK one)
+        while True:   # Get the SCTP layer   
+          tul = type(ul)
+          if tul == SCTP:
```
Comment (p-l-) [inline scapy/layers/sctp.py]: This should be written as `if isinstance(ul, SCTP):`. Also, `tul` is not needed.

File: scapy/layers/sctp.py [hunk 11]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
                     XShortField("stream_id", None),
                     XShortField("stream_seq", None),
-                    XIntField("proto_id", None),
+                    ProtoIdField("proto_id", None, SCTP_PAYLOAD_PROTOCOL_INDENTIFIERS),
                     PadField(StrLenField("data", None, length_from=lambda pkt: pkt.len-16),
                              4, padwith=b"\x00"),
                     ]
+    
+    def guess_payload_class(self, payload):
+      protoClassname = {  # Dict used only if self.proto_id is defined
```
Comment (p-l-) [inline scapy/layers/sctp.py]: This is not needed and should be handled by `bind_layer()` calls.

File: scapy/layers/sctp.py [hunk 12]
```diff
@@ -425,21 +425,114 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
+        ul = self.underlayer  # Not always the SCTP layer (may be the SACK one)
+        while True:   # Get the SCTP layer   
+          tul = type(ul)
+          if tul == SCTP:
+            break
+          elif tul == type(None): # Should not reach this point but just in case ...
+            warning ('Could not find the SCTP layer.')
+            return conf.raw_layer
+          ul = ul.underlayer
+        self.fields['dport'] = ul.dport; self.dport = ul.dport
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Are you setting fields that are not declared? I don't get it.

File: scapy/layers/sctp.py [hunk 13]
```diff
@@ -436,11 +499,15 @@ class SCTPChunkData(_SCTPChunkGuessPayload, Packet):
                     XIntField("tsn", None),
                     XShortField("stream_id", None),
                     XShortField("stream_seq", None),
-                    XIntField("proto_id", None),
+                    IntEnumField("proto_id", None, SCTP_PAYLOAD_PROTOCOL_INDENTIFIERS),
                     PadField(StrLenField("data", None, length_from=lambda pkt: pkt.len-16),
                              4, padwith=b"\x00"),
                     ]
 
+    def default_payload_class(self, payload):
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Why is that needed?

File: scapy/layers/sctp.py [hunk 14]
```diff
@@ -425,7 +425,70 @@ class SCTPChunkParamAdaptationLayer(_SCTPChunkParam, Packet):
...
+	52: 'WebRTC Binary Partial',
+	53: 'WebRTC Binary',
+	54: 'WebRTC String Partial',
+	55: '3GPP PUA',
+	56: 'WebRTC String Empty',
+	57: 'WebRTC Binary Empty'	
+}
+
 class SCTPChunkData(_SCTPChunkGuessPayload, Packet):
+# TODO : add a padding function in post build if this layer is used to generate SCTP chunk data
```
Comment (p-l-) [inline scapy/layers/sctp.py]: Can you indent this comment?
